### ISSUE TITLE
[Xamarin.Android] HotReload with SQL Lite 

### ISSUE THREAD

**Issue by Suprndm** [adospace/reactorui-xamarin#13]
Hi,

I have a basic Sql lite database with one table and one int column. I followed this tutorial https://docs.microsoft.com/fr-fr/xamarin/xamarin-forms/data-cloud/data/databases

Sadly, as soon as I press hotreload, executing db queries fires an exception. The connection to the database does not seem to work anymore. The exception is : ""Specified cast is not valid." and seems very "deep" (mono and mscorlib in stacktrace). Even if the connection is re instanciated after the hotreload, the connection.TableMappings are preserved, already populated.

I don't use any state preservation. All the app is re instanciated, re constructed, even the SQL Lite connection. But some how it seems to preserve old table mappings from the previous run, and fails on the next queries.

Restarting completly the app solves the issue.

I'm still investigating the issue. 
Using EFCore.SQL lite did not produce this issue. I gave up on EFCore on xamarin because of horrible startup performances.
So i'm using bear sql lite.

Do you know what exactly happens with runtime objects when hotreloading vs restarting the app ? 
How could the connection still have those table mappings even after re constructing the object.

Thanks

**Comment by adospace** [comment]
Hi, can you share some of the component code ( just to understand how you use the connection to sqlite)? Usually the connection to db should be passed using Context down to the component tree (I.e. not in component state)

**Comment by Suprndm** [comment]
Mmmh it's gonna be hard to show you in a simple way... I only use reactorUI for hotreload. So i  have one main component, then i use SkiaSharp as a render engine, and all my app is drawn like that. Not the typical usecase i guess. But its very powerful that way.

Anyway 
my main component is :  
`public class OdinComponent : RxComponent<GameRoot>`

GameRoot then instantiate sql lite in a straightforward way :  
`Database = new SQLiteConnectionWithLock(new SQLiteConnectionString(Constants.DatabasePath));`

Looks like the connection is passed via component state...

So basically i should try to remove the connection from the component state, and keep it direclty in the main component.



On a side note, I just tried an SQL Lite alternative called Realm. And when I hotreload, i get an exception as well  hahaha,
System.NotSupportedException: 'AddDefaultTypes should be called before creating a Realm instance with the default schema. If you see this error, please report it to help@realm.io.'

Looks like every one is not quite used to develop libraries that are "hot reload ready" just yet.

**Comment by Suprndm** [comment]
I removed the SQLConnection from the main component state. And put it directly as a member of the main component.
Still get the exception.
Here is the line that fails after hotreload.

`      var activity = Database.Query<Activity>("SELECT * FROM Activity").FirstOrDefault();`

When i change Activity to Object. It works, i mean i get the correct quantity of items. Just not in the good type, and i cant cast it afterwards.
Could it be that  after the hotreload, the dll changes in a way that the connection does not recognize anymore the Activity type as the one it got in its tableMapping...

**Comment by adospace** [comment]
Ok, yes, the SqliteConnection should not be a component state property (much probably that's the problem).
You have 2 alternatives:
1) Open the connection, execute commands/queries, finally close it. I suggest putting everything inside a `using` block and with the async/await calls
2) Open a connection to SQLite at startup and put it in a Context (https://adospace.gitbook.io/reactorui/guide/components-with-a-state/component-context)

**Comment by adospace** [comment]
Just see your comment now, the connection should not be in the state nor as component property (like a props) but please try as explained above:
1) First option:
```
using (var db = CreateDatabase(...))
{
var activity = await db.Query<Activity>("SELECT * FROM Activity").FirstOrDefaultAsync();
}
```
2) second option:
```
_app = RxApplication.Create<ShellComponent>(this)
    .WithContext("Database", CreateSingletonDatabaseInstance())
```
and in component:
```
var database = (SqlteConnection)Context["Database"];
var activity = database .Query<Activity>("SELECT * FROM Activity").FirstOrDefault();
```

**Comment by Suprndm** [comment]
Okay, i was trying to create a simple project to reproduce the issue, and I cloned the reactorUI  sample. That's where I saw you created a separeted project for services. It reminds me I no longer needed to have my data Model and DB connection in the main project. So I created a separed project for them. And now i no longer Have the issue !
I suspect that the dlls that contains the DB Model and connection are not concerned any more by hotreload, and so the cast issue is no more.

For science i'm gonna try your solutions. But if that was a pure dll change issue, it should not fix it. Let's see !

**Comment by Suprndm** [comment]
wooow okay. So. I think my issue has nothing to do with SQL Lite afterall.
I tried your second option : 
`
            _rxApp = RxApplication.Create<FooComponent>(this)
                .WithContext("Database", new FooDatabase())
`

Then in the main component 
`  var database = (FooDatabase)Context["Database"];`

It works fine when starting the app. But at the first hot reload, i got
System.InvalidCastException: Specified cast is not valid.

I did not even added the code that query the DB. just the cast line produce the exception ! Same exception i got when using connection with SQL lite. 
Truth is SQL lite is casting to model the results of the query. So there is an issue with casting types from the dlls that are hot reloaded.

I reproduced the issue even simplier
Create a simple class in the xamarin project.
Mine is Activity.
Put it in AppContext 
         `  .WithContext("Activity", new Activity())`
Cast it in component  
              `  var activity = (Activity)Context["Activity"];`
 Get the InvalidCastException exception.

Somehow the program cannot cast an object that was instanciated using an old dll , with the newer dll ?

I know you would recommand to put all business and persistance logic outside of the xamarin project that is hot reloaded, but did you know this issue ?
